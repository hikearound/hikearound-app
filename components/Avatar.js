import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { TouchableOpacity, Image } from 'react-native';
import * as Permissions from 'expo-permissions';
import * as ImagePicker from 'expo-image-picker';
import { opacities } from '../constants/Index';
import { updateAvatar } from '../actions/User';
import shrinkImageAsync from '../utils/User';

const propTypes = {
    dispatchAvatar: PropTypes.func.isRequired,
    avatar: PropTypes.string.isRequired,
    size: PropTypes.number,
};

const defaultProps = {
    size: 60,
};

function mapStateToProps(state) {
    return {
        avatar: state.userReducer.avatar,
    };
}

function mapDispatchToProps(dispatch) {
    return {
        dispatchAvatar: (photoData) => dispatch(updateAvatar(photoData)),
    };
}

class Avatar extends React.Component {
    checkPhotoPermissions = async () => {
        const { status } = await Permissions.getAsync(Permissions.CAMERA_ROLL);
        if (status !== 'granted') {
            const { newStatus } = await Permissions.askAsync(
                Permissions.CAMERA_ROLL,
            );
            if (newStatus === 'granted') {
                this.launchPhotoPicker();
            }
        } else {
            this.launchPhotoPicker();
        }
    };

    launchPhotoPicker = async () => {
        const photo = await ImagePicker.launchImageLibraryAsync();
        if (!photo.cancelled) {
            this.uploadImage(photo.uri);
        }
    };

    uploadImage = async (originalUri) => {
        const { dispatchAvatar } = this.props;

        const { uri } = await shrinkImageAsync(originalUri);
        const response = await fetch(uri);
        const blob = await response.blob();

        const photoData = {
            uri,
            blob,
        };

        dispatchAvatar(photoData);
    };

    render() {
        const { avatar, size } = this.props;
        return (
            <TouchableOpacity
                onPress={() => {
                    this.checkPhotoPermissions();
                }}
                activeOpacity={opacities.regular}
            >
                <Image
                    source={{ uri: avatar }}
                    style={{
                        height: size,
                        width: size,
                        borderRadius: size / 2,
                    }}
                />
            </TouchableOpacity>
        );
    }
}

Avatar.propTypes = propTypes;
Avatar.defaultProps = defaultProps;

export default connect(
    mapStateToProps,
    mapDispatchToProps,
)(Avatar);
