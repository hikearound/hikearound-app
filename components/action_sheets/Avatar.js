import { ActionSheetIOS } from 'react-native';
import {
    checkImageLibraryPermissions,
    checkCameraPermissions,
} from '@utils/PhotoPicker';

const cancelButtonIndex = 2;

export function getSheetOptions(t) {
    return [
        t('sheet.avatar.item.camera'),
        t('sheet.avatar.item.library'),
        t('label.common.cancel'),
    ];
}

export function avatarActionSheet(t, dispatchAvatar) {
    const options = getSheetOptions(t);

    ActionSheetIOS.showActionSheetWithOptions(
        {
            options,
            cancelButtonIndex,
        },

        (buttonIndex) => {
            if (buttonIndex === 0) {
                checkCameraPermissions(dispatchAvatar);
            } else if (buttonIndex === 1) {
                checkImageLibraryPermissions(dispatchAvatar);
            }
        },
    );
}
