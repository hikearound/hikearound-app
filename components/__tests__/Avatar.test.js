import React from 'react';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import renderer from 'react-test-renderer';
import Avatar from '../Avatar';

const mockStore = configureMockStore();
const store = mockStore({
    userReducer: {
        avatar: 'https://patdugan.me/images/me.jpg',
    },
});

test('Avatar renders correctly', () => {
    const tree = renderer
        .create(
            <Provider store={store}>
                <Avatar />
            </Provider>,
        )
        .toJSON();
    expect(tree).toMatchSnapshot();
});
