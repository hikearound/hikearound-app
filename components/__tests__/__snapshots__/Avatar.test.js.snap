// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar renders correctly 1`] = `
<View>
  <View
    style={
      Object {
        "backgroundColor": "#DADADA",
        "borderRadius": 30,
        "height": 60,
        "width": 60,
      }
    }
  />
</View>
`;
